src/rust/inetstack/protocols/tcp/peer.rs

```diff
@@ -295,196 +339,211 @@ impl TcpPeer {
+        // TODO: Currently we do not handle close correctly because we continue to receive packets at this point to finish the TCP close protocol.
+        // 1. We do not remove the endpoint from the addresses table
+        // 2. We do not remove the queue from the queue table.
+        // As a result, we have stale closed queues that are labelled as closing. We should clean these up.
+        // look up socket
+        let addr: SocketAddrV4 = match inner.qtable.borrow_mut().get_mut(&qd) {
+            Some(InetQueue::Tcp(queue)) => {
+                match queue.get_socket() {
+                    // Closing an active socket
+                    Socket::Established(socket) => {
+                        socket.close()?;
+                        queue.set_socket(Socket::Closing(socket.clone()));
+                        return Ok(());
+                    },
+                    Socket::Listening(socket) => socket.endpoint(),
+                    Socket::Inactive(Some(addr)) => addr.clone(),
+                    _ => return Err(Fail::new(libc::ENOTSUP, "close not implemented for listening sockets")),
                 }
             },
-
-            Some(..) => return Err(Fail::new(ENOTSUP, "close not implemented for listening sockets")),
-            None => return Err(Fail::new(EBADF, "bad queue descriptor")),
-        }
-
-        Ok(())
+            _ => return Err(Fail::new(libc::EBADF, "bad queue descriptor")),
+        };
+        inner.addresses.remove(&SocketId::Passive(addr));
+        Err(Fail::new(libc::ENOTSUP, "close not implemented for listening sockets"))
     }
 
-    pub fn remote_mss(&self, fd: QDesc) -> Result<usize, Fail> {
+    pub fn remote_mss(&self, qd: QDesc) -> Result<usize, Fail> {
+        #[cfg(feature = "profiler")]
+        timer!("tcp::remote_mss");
         let inner = self.inner.borrow();
-        let key = match inner.sockets.get(&fd) {
-            Some(Socket::Established { local, remote }) => (*local, *remote),
-            Some(..) => return Err(Fail::new(ENOTCONN, "connection not established")),
-            None => return Err(Fail::new(EBADF, "bad queue descriptor")),
-        };
-        match inner.established.get(&key) {
-            Some(ref s) => Ok(s.remote_mss()),
-            None => Err(Fail::new(ENOTCONN, "connection not established")),
+        let qtable: Ref<IoQueueTable<InetQueue>> = inner.qtable.borrow();
+        match qtable.get(&qd) {
+            Some(InetQueue::Tcp(queue)) => match queue.get_socket() {
+                Socket::Established(socket) => Ok(socket.remote_mss()),
+                _ => Err(Fail::new(libc::ENOTCONN, "connection not established")),
+            },
+            _ => Err(Fail::new(libc::EBADF, "bad queue descriptor")),
         }
     }
 
-    pub fn current_rto(&self, fd: QDesc) -> Result<Duration, Fail> {
+    pub fn current_rto(&self, qd: QDesc) -> Result<Duration, Fail> {
+        #[cfg(feature = "profiler")]
+        timer!("tcp::current_rto");
```

Comment:
Could we add changes that concern performance observation in a different commit?